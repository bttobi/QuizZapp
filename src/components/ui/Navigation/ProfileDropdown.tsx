import React from 'react';
import {
  Avatar,
  Dropdown,
  DropdownItem,
  DropdownMenu,
  DropdownTrigger,
  Spinner,
} from '@nextui-org/react';
import { useAuthUser } from 'react-auth-kit';
import messages from '../../../api/messages/messages.json';
import {
  useGetUserRanking,
  useSignOutUser,
} from '../../../api/hooks/user.hooks';
import { getRankingColor } from '../../../helpers/rankingStyling';
import { useNavigate } from 'react-router-dom';

const ProfileDropdown = () => {
  const { signOut } = useSignOutUser();
  const { rankingData, isFetching } = useGetUserRanking();
  const navigate = useNavigate();
  const auth = useAuthUser();
  const user = auth();

  return (
    <Dropdown placement="bottom-end">
      <DropdownTrigger>
        <Avatar
          isBordered
          as="button"
          className="transition-transform"
          color="secondary"
          name={user?.email.slice(0, 2).toUpperCase()}
          size="sm"
          src=""
        />
      </DropdownTrigger>
      <DropdownMenu
        color="primary"
        aria-label="Profile Actions"
        variant="solid"
      >
        <DropdownItem
          key="user-details"
          className="h-14 gap-2 pointer-events-none"
          textValue={user?.email}
        >
          <p className="font-semibold">{messages.signedAs}</p>
          <p className="font-semibold">{user?.email}</p>
        </DropdownItem>
        <DropdownItem
          key="ranking"
          textValue={`${messages.points}: ${rankingData?.points}`}
          className="pointer-events-none"
        >
          {isFetching ? (
            <Spinner data-testid="spinner" color="white" />
          ) : (
            <div className="flex flex-col font-semibold">
              <span
                className={getRankingColor(rankingData?.place || 0)}
              >{`${messages.place}: ${rankingData?.place}`}</span>
              <span className="text-secondary">{`${messages.points}: ${rankingData?.points}`}</span>
            </div>
          )}
        </DropdownItem>
        <DropdownItem
          key="profile"
          textValue={`${messages.myProfile}`}
          color="primary"
          onClick={() => navigate(`/profile/${user?.userID}`)}
        >
          {messages.myProfile}
        </DropdownItem>
        {/* TODO: possibly add user profile <DropdownItem key="settings">My settings</DropdownItem> */}
        <DropdownItem onClick={() => signOut()} key="signout" color="danger">
          {messages.signOut}
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
  );
};

export default ProfileDropdown;
