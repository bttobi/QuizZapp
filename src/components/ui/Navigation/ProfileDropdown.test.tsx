import React from 'react';
import {
  AllTheProviders,
  fireEvent,
  render,
  renderHook,
  waitFor,
} from '../../../tests/test-utils';
import ProfileDropdown from './ProfileDropdown';
import { useSignIn } from 'react-auth-kit';

const setUser = renderHook(() => useSignIn(), { wrapper: AllTheProviders });
const setup = () => {
  setUser.result?.current?.({
    token: 'token',
    tokenType: 'Bearer',
    expiresIn: 420,
    refreshToken: 'token',
    refreshTokenExpireIn: 1337,
    authState: {
      email: 'test@gmail.com',
      userID: 1,
      token: 'token',
    },
  });
  return render(<ProfileDropdown />);
};

describe('ProfileDropdown component', () => {
  it('renders the component', async () => {
    const { container } = setup();
    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('displays all the user info', async () => {
    const { container, getByText } = setup();
    await waitFor(() => {
      expect(container.querySelector('[data-slot]')).toBeInTheDocument();
    });
    fireEvent.click(container.querySelector('[data-slot]') as HTMLElement);
    await waitFor(() => {
      expect(getByText('Signed in as')).toBeInTheDocument();
    });
  });

  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
