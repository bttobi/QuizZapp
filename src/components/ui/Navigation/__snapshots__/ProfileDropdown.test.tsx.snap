// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileDropdown component > matches snapshot 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="flex relative justify-center items-center box-border overflow-hidden align-middle outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-8 h-8 text-tiny bg-secondary text-secondary-foreground rounded-full ring-2 ring-offset-2 ring-offset-background dark:ring-offset-background-dark ring-secondary z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased transition-transform"
    data-slot="trigger"
    id="react-aria-:rl:"
    type="button"
  >
    <span
      aria-label="TE"
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 font-normal text-center text-inherit"
      role="img"
    >
      TE
    </span>
  </button>
  <div
    class="Toastify"
  />
</div>
`;
